### PR TITLE
Write path and hacky read path

### PR DESCRIPTION
**Pull request by ChinmaySKulkarni**
- Added Expression for BLOBs
- Added metadata code and SERDES for the same
- Integrated with LobStore APIs
- Added hacky read path and support for rs.getBinaryStream()

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BlobTypeIT.java [hunk 1]
```diff
@@ -38,19 +41,30 @@ public static synchronized void doSetup() throws Exception {
...
+                prepStmt.execute();
+            }
+            conn.commit();
+        }
+        try (Connection conn = DriverManager.getConnection(getUrl());
+                Statement stmt = conn.createStatement()) {
+            ResultSet rs = stmt.executeQuery("SELECT image FROM " + tableName);
+            assertTrue(rs.next());
+            InputStream inputStr = rs.getBinaryStream(1);
+            assertNotNull(inputStr);
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/BlobTypeIT.java]: I'll modify this to do rs.getBlob and check the PheonixBlob once https://github.com/dbwong/phoenix/pull/5 is merged.

File: phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixResultSet.java [hunk 2]
```diff
@@ -316,16 +317,24 @@ public BigDecimal getBigDecimal(String columnLabel, int scale) throws SQLExcepti
...
     @Override
     public InputStream getBinaryStream(String columnLabel) throws SQLException {
-        throw new SQLFeatureNotSupportedException();
+        return getBinaryStream(findColumn(columnLabel));
     }
 
     @Override
     public Blob getBlob(int columnIndex) throws SQLException {
+        // TODO: Make a PhoenixBlob representation and return that here
         throw new SQLFeatureNotSupportedException();
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixResultSet.java]: Will be able to use what's in https://github.com/dbwong/phoenix/pull/5

File: phoenix-core/src/main/java/org/apache/phoenix/schema/types/PBlob.java [hunk 3]
```diff
@@ -17,15 +17,44 @@
...
+        BlobExpression.BlobMetaData blobMetaData;
+        try {
+            blobMetaData = BlobExpression.BlobMetaData.deserializeBlobMetaData(bytesCopy);
+        } catch (IOException | ClassNotFoundException e) {
+            throw new RuntimeException("Could not deserialize BLOB metadata");
+        }
+        if (blobMetaData != null) {
+            BlobExpression dummyBlobExpr = new BlobExpression(null, null, null);
+            // TODO: Should return a PhoenixBlob object rather than an InputStream
+            return dummyBlobExpr.getInputStream(blobMetaData);
```
Comment (ChinmaySKulkarni) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/types/PBlob.java]: Same here. Will use PhoenixBlob after https://github.com/dbwong/phoenix/pull/5 is merged

File: phoenix-core/src/main/java/org/apache/phoenix/expression/BlobExpression.java [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    BlobMetaData getBlobMetaData(boolean onDisk, String lobLocation, long lastUpdatedTime,
+            SUPPORTED_FORMATS format, UUID versionNo) {
+        return new BlobMetaData(onDisk, lobLocation, lastUpdatedTime, format, versionNo);
+    }
+
+    /**
+     * Stores metadata about the BLOB
+     */
+    public static final class BlobMetaData implements Serializable {
+
```
Comment (dbwong) [inline phoenix-core/src/main/java/org/apache/phoenix/expression/BlobExpression.java:106]: Given the blob api prestoring a size here would be nice.

Comment (ChinmaySKulkarni) [reply phoenix-core/src/main/java/org/apache/phoenix/expression/BlobExpression.java:106]: Will do.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BlobTypeIT.java [hunk 5]
```diff
@@ -63,98 +41,32 @@ public static synchronized void doSetup() throws Exception {
...
+            try (PreparedStatement prepStmt = conn.prepareStatement(upsertQuery)) {
+                prepStmt.setString(1, "ID1");
+                prepStmt.setBlob(2, input);
+                prepStmt.execute();
+            }
+            conn.commit();
+        }
+        try (Connection conn = DriverManager.getConnection(getUrl());
+                Statement stmt = conn.createStatement()) {
+            TestUtil.printResultSet(stmt.executeQuery("SELECT image FROM " + tableName ));
```
Comment (dbwong) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/BlobTypeIT.java:63]: Curious if this output is reasonable on the blob column?

**Review by dbwong — APPROVED** [review]

**Pull request merged**